diff --git a/mlir/lib/Target/LLVM/CMakeLists.txt b/mlir/lib/Target/LLVM/CMakeLists.txt
index a6f1436e1d3a..7ea34ca1a3be 100644
--- a/mlir/lib/Target/LLVM/CMakeLists.txt
+++ b/mlir/lib/Target/LLVM/CMakeLists.txt
@@ -243,6 +243,10 @@ if (OCLOC_FOUND)
     PRIVATE
     MLIR_XEVM_OCLOC_LIB_AVAILABLE=1
   )
+
+  if(LLVM_BUILD_LLVM_DYLIB)
+    set_property(GLOBAL APPEND PROPERTY MLIR_SHARED_LIBS ${OCLOC_LIBRARY})
+  endif()
 else()
   target_compile_definitions(obj.MLIRXeVMTarget
     PRIVATE
diff --git a/mlir/tools/mlir-shlib/CMakeLists.txt b/mlir/tools/mlir-shlib/CMakeLists.txt
index a33c70c5807b..08c26b734845 100644
--- a/mlir/tools/mlir-shlib/CMakeLists.txt
+++ b/mlir/tools/mlir-shlib/CMakeLists.txt
@@ -44,7 +44,9 @@ if(LLVM_BUILD_LLVM_DYLIB)
     LINK_COMPONENTS
     ${mlir_llvm_link_components}
   )
-  target_link_libraries(MLIR PRIVATE ${LLVM_PTHREAD_LIB})
+
+  get_property(mlir_slibs GLOBAL PROPERTY MLIR_SHARED_LIBS)
+  target_link_libraries(MLIR PRIVATE ${LLVM_PTHREAD_LIB} ${mlir_slibs})
 endif()
 
 #message("Libraries included in libMLIR.so: ${mlir_libs}")
